File: src/Fountain.cpp
#include "Fountain.h"

#include <QMap>
#include <QRegularExpression>
#include <QSet>
#include <algorithm>

namespace {

const QRegularExpression sceneHeadingPattern(
    R"(^\s*\.?\s*(INT|EXT|EST|INT\/EXT|INT\.\/EXT|I\/E)[\.\s])",
    QRegularExpression::CaseInsensitiveOption);

const QRegularExpression likelySceneHeadingPattern(
    R"(^\s*\.?\s*(IN|ITN|EX|EXR|ENT|EXT|INT)[\.\s-])",
    QRegularExpression::CaseInsensitiveOption);

const QRegularExpression titleKeyPattern(R"(^[A-Za-z][A-Za-z0-9 _-]*:\s*.+$)");

}

FountainDocument FountainParser::parse(const QString &text) const
{
    FountainDocument document;
    const QStringList rawLines = text.split('\n');
    QMap<QString, CharacterStats> characterMap;
    QMap<QString, QSet<int>> characterScenes;

    bool inTitlePage = true;
    bool expectingDialogue = false;
    QString activeCharacter;
    int currentScene = -1;
    int offset = 0;
    int openNotes = 0;

    for (int i = 0; i < rawLines.size(); ++i) {
        const QString lineText = rawLines.at(i);
        const QString trimmed = lineText.trimmed();
        FountainLine line;
        line.index = i;
        line.start = offset;
        line.length = lineText.length();
        line.text = lineText;

        openNotes += lineText.count(QStringLiteral("[["));
        openNotes -= lineText.count(QStringLiteral("]]"));

        if (trimmed.isEmpty()) {
            line.type = FountainLineType::Empty;
            expectingDialogue = false;
            activeCharacter.clear();
        } else if (inTitlePage && titleKeyPattern.match(trimmed).hasMatch()) {
            line.type = FountainLineType::TitlePage;
        } else {
            inTitlePage = false;

            if (trimmed.startsWith(QStringLiteral("/*")) || trimmed.startsWith(QStringLiteral("*/"))
                || (trimmed.startsWith(QStringLiteral("[[")) && trimmed.endsWith(QStringLiteral("]]")))) {
                line.type = FountainLineType::Note;
            } else if (trimmed == QStringLiteral("===")) {
                line.type = FountainLineType::PageBreak;
            } else if (trimmed.startsWith('#')) {
                line.type = FountainLineType::Section;
            } else if (trimmed.startsWith('=')) {
                line.type = FountainLineType::Synopsis;
            } else if (trimmed.startsWith('~')) {
                line.type = FountainLineType::Lyric;
            } else if (trimmed.startsWith('>') && trimmed.endsWith('<')) {
                line.type = FountainLineType::Centered;
            } else if (isSceneHeading(trimmed)) {
                line.type = FountainLineType::SceneHeading;
                ++document.sceneCount;
                currentScene = document.sceneCount;
                expectingDialogue = false;
                activeCharacter.clear();
            } else if (isTransition(trimmed)) {
                line.type = FountainLineType::Transition;
                expectingDialogue = false;
                activeCharacter.clear();
            } else if (isCharacterCue(trimmed)) {
                line.type = FountainLineType::Character;
                activeCharacter = normalizedCharacterName(trimmed);
                line.character = activeCharacter;
                expectingDialogue = true;
                CharacterStats &stats = characterMap[activeCharacter];
                stats.name = activeCharacter;
                if (currentScene >= 0) {
                    characterScenes[activeCharacter].insert(currentScene);
                }
            } else if (expectingDialogue && trimmed.startsWith('(') && trimmed.endsWith(')')) {
                line.type = FountainLineType::Parenthetical;
                line.character = activeCharacter;
            } else if (expectingDialogue) {
                line.type = FountainLineType::Dialogue;
                line.character = activeCharacter;
                const int words = wordCount(trimmed);
                document.wordCount += words;
                CharacterStats &stats = characterMap[activeCharacter];
                stats.name = activeCharacter;
                stats.dialogueLines += 1;
                stats.dialogueWords += words;
                if (stats.sampleLines.size() < 5) {
                    stats.sampleLines.push_back(trimmed);
                }
            } else {
                line.type = FountainLineType::Action;
                document.wordCount += wordCount(trimmed);
            }
        }

        if (line.type == FountainLineType::Action && isLikelySceneHeading(trimmed)) {
            FountainDiagnostic diagnostic;
            diagnostic.line = i;
            diagnostic.start = line.start;
            diagnostic.length = line.length;
            diagnostic.message = QStringLiteral("Possible scene heading typo or missing INT./EXT. prefix.");
            diagnostic.replacement = trimmed.toUpper();
            if (diagnostic.replacement.startsWith(QStringLiteral("IN "))) {
                diagnostic.replacement.replace(0, 2, QStringLiteral("INT."));
            } else if (diagnostic.replacement.startsWith(QStringLiteral("EX "))) {
                diagnostic.replacement.replace(0, 2, QStringLiteral("EXT."));
            }
            document.diagnostics.push_back(diagnostic);
        }

        if (line.type == FountainLineType::Action && trimmed.size() > 2 && trimmed == trimmed.toUpper()
            && !trimmed.endsWith('.') && !trimmed.contains(':')) {
            FountainDiagnostic diagnostic;
            diagnostic.line = i;
            diagnostic.start = line.start;
            diagnostic.length = line.length;
            diagnostic.message = QStringLiteral("Uppercase action may be intended as a character cue.");
            diagnostic.replacement = trimmed;
            document.diagnostics.push_back(diagnostic);
        }

        if (line.type == FountainLineType::Transition && !trimmed.endsWith(':')) {
            FountainDiagnostic diagnostic;
            diagnostic.line = i;
            diagnostic.start = line.start;
            diagnostic.length = line.length;
            diagnostic.message = QStringLiteral("Transitions conventionally end with a colon.");
            diagnostic.replacement = trimmed + QStringLiteral(":");
            document.diagnostics.push_back(diagnostic);
        }

        document.lines.push_back(line);
        offset += lineText.length() + 1;
    }

    if (openNotes > 0) {
        FountainDiagnostic diagnostic;
        diagnostic.line = qMax(0, rawLines.size() - 1);
        diagnostic.start = qMax(0, text.length() - 1);
        diagnostic.length = 0;
        diagnostic.message = QStringLiteral("A note appears to be missing a closing ]].");
        diagnostic.replacement = QStringLiteral("]]");
        document.diagnostics.push_back(diagnostic);
    }

    for (auto it = characterMap.begin(); it != characterMap.end(); ++it) {
        CharacterStats stats = it.value();
        stats.sceneCount = characterScenes.value(it.key()).size();
        stats.estimatedSeconds = (stats.dialogueWords * 0.38) + (stats.sceneCount * 8.0);
        document.characters.push_back(stats);
    }

    std::sort(document.characters.begin(), document.characters.end(),
              [](const CharacterStats &left, const CharacterStats &right) {
                  if (left.estimatedSeconds == right.estimatedSeconds) {
                      return left.name < right.name;
                  }
                  return left.estimatedSeconds > right.estimatedSeconds;
              });

    return document;
}

QString FountainParser::normalizedCharacterName(QString cue)
{
    cue = cue.trimmed();
    cue.remove('^');
    cue.remove(QRegularExpression(R"(\s*\(.*\)\s*$)"));
    return cue.trimmed();
}

bool FountainParser::isSceneHeading(const QString &trimmed)
{
    return sceneHeadingPattern.match(trimmed).hasMatch();
}

bool FountainParser::isLikelySceneHeading(const QString &trimmed)
{
    return likelySceneHeadingPattern.match(trimmed).hasMatch();
}

bool FountainParser::isTransition(const QString &trimmed)
{
    static const QSet<QString> transitions = {
        QStringLiteral("CUT TO"),
        QStringLiteral("CUT TO:"),
        QStringLiteral("FADE IN"),
        QStringLiteral("FADE IN:"),
        QStringLiteral("FADE OUT"),
        QStringLiteral("FADE OUT:"),
        QStringLiteral("DISSOLVE TO"),
        QStringLiteral("DISSOLVE TO:"),
        QStringLiteral("SMASH CUT TO"),
        QStringLiteral("SMASH CUT TO:"),
        QStringLiteral("MATCH CUT TO"),
        QStringLiteral("MATCH CUT TO:")
    };
    const QString upper = trimmed.toUpper();
    return upper.endsWith(QStringLiteral(" TO:")) || transitions.contains(upper);
}

bool FountainParser::isCharacterCue(const QString &trimmed)
{
    if (trimmed.length() < 2 || trimmed.length() > 42) {
        return false;
    }
    if (trimmed.contains(':') || trimmed.endsWith('.') || trimmed.startsWith('>')) {
        return false;
    }
    const QString withoutDual = trimmed.endsWith('^') ? trimmed.left(trimmed.length() - 1).trimmed() : trimmed;
    if (withoutDual != withoutDual.toUpper()) {
        return false;
    }
    return withoutDual.contains(QRegularExpression(R"([A-Z0-9])"));
}

int FountainParser::wordCount(const QString &text)
{
    static const QRegularExpression words(R"(\b[\p{L}\p{N}']+\b)");
    int count = 0;
    auto match = words.globalMatch(text);
    while (match.hasNext()) {
        match.next();
        ++count;
    }
    return count;
}

File: src/fountain.ts
import type { CharacterStats, FountainDocument, FountainLine, LineType } from './types';

const scenePattern = /^(INT|EXT|EST|INT\/EXT|INT\.\/EXT|I\/E)(?:[.\s-]|$)/i;
const likelyScenePattern = /^(IN|ITN|EX|EXR|ENT|EXT|INT)[.\s-]/i;
const titlePattern = /^([A-Za-z][A-Za-z0-9 _-]*):(?:\s*(.*))?$/;
const transitions = new Set(['CUT TO', 'CUT TO:', 'FADE IN', 'FADE IN:', 'FADE OUT', 'FADE OUT:', 'DISSOLVE TO', 'DISSOLVE TO:', 'SMASH CUT TO', 'SMASH CUT TO:', 'MATCH CUT TO', 'MATCH CUT TO:']);
const countWords = (value: string) => value.match(/[\p{L}\p{N}’']+/gu)?.length ?? 0;

export function unescapeFountain(value: string): string {
  return value.replace(/\\([*_@!~^#.<>\[\]\\])/g, '$1');
}

export function stripInlineNotes(value: string): string {
  return value.replace(/\[\[[\s\S]*?\]\]/g, '').replace(/\[\[[\s\S]*$/g, '').replace(/^[\s\S]*?\]\]/, '').trimEnd();
}

export function printableMarkup(line: FountainLine): string {
  let value = line.type === 'action' ? line.text.trimEnd() : line.text.trim();
  if (line.forced && ['scene', 'character', 'action', 'transition'].includes(line.type)) value = line.text.trim().slice(1).trimStart();
  if (line.type === 'scene') value = value.replace(/\s+#([^#]+)#\s*$/, '').trimEnd();
  if (line.type === 'character') value = value.replace(/\^\s*$/, '').trimEnd();
  if (line.type === 'lyric') value = value.replace(/^~/, '').trimStart();
  if (line.type === 'centered') value = value.replace(/^>\s*/, '').replace(/\s*<$/, '');
  return stripInlineNotes(value);
}

export function printableText(line: FountainLine): string {
  return unescapeFountain(printableMarkup(line));
}

function isTransition(value: string) {
  const upper = value.toUpperCase();
  return upper.endsWith(' TO:') || transitions.has(upper);
}

function normalizeCharacter(value: string) {
  return value.replace(/^@/, '').replace(/\^\s*$/, '').replace(/\s*\(.*\)\s*$/, '').trim();
}

function isNaturalCharacter(value: string, previousBlank: boolean, nextLine: string) {
  if (!previousBlank || !nextLine.trim() || value.length < 2 || value.length > 42 || value.includes(':') || value.endsWith('.') || value.startsWith('>')) return false;
  const withoutDual = value.replace(/\^\s*$/, '').trim();
  return withoutDual === withoutDual.toUpperCase() && /[A-Z0-9]/.test(withoutDual);
}

function classifyTitlePage(rawLines: string[]) {
  const titleLines = new Set<number>();
  const titlePage: Record<string, string> = {};
  let activeKey = '', sawKey = false;
  for (let index = 0; index < rawLines.length; index++) {
    const raw = rawLines[index];
    const match = raw.trim().match(titlePattern);
    if (match) {
      sawKey = true; activeKey = match[1].toLowerCase(); titleLines.add(index);
      titlePage[activeKey] = match[2]?.trim() || '';
      continue;
    }
    if (sawKey && activeKey && /^[ \t]+\S/.test(raw)) {
      titleLines.add(index); titlePage[activeKey] = [titlePage[activeKey], raw.trim()].filter(Boolean).join('\n'); continue;
    }
    if (!raw.trim() && sawKey) { titleLines.add(index); continue; }
    break;
  }
  return { titleLines, titlePage };
}

export function parseFountain(text: string): FountainDocument {
  const rawLines = text.replace(/\r\n?/g, '\n').split('\n');
  const title = classifyTitlePage(rawLines);
  const document: FountainDocument = { lines: [], characters: [], diagnostics: [], sceneCount: 0, wordCount: 0, titlePage: title.titlePage };
  const characterMap = new Map<string, CharacterStats>();
  const characterScenes = new Map<string, Set<number>>();
  let expectingDialogue = false, activeCharacter = '', activeDual = false, currentScene = -1, offset = 0, inBoneyard = false, noteDepth = 0;

  rawLines.forEach((lineText, index) => {
    const trimmed = lineText.trim();
    const previousBlank = index === 0 || !rawLines[index - 1].trim();
    const nextLine = rawLines[index + 1] || '';
    const line: FountainLine = { index, start: offset, length: lineText.length, text: lineText, type: 'action' };
    const startsBoneyard = !inBoneyard && lineText.includes('/*');
    const endsBoneyard = lineText.includes('*/');

    if (inBoneyard || startsBoneyard) {
      line.type = 'boneyard'; inBoneyard = !endsBoneyard; expectingDialogue = false; activeCharacter = '';
    } else if (title.titleLines.has(index)) {
      line.type = trimmed ? 'title' : 'empty';
    } else if (!trimmed) {
      line.type = 'empty'; expectingDialogue = false; activeCharacter = ''; activeDual = false;
    } else if (/^={3,}$/.test(trimmed)) {
      line.type = 'page-break'; expectingDialogue = false; activeCharacter = '';
    } else if (noteDepth > 0 || trimmed.startsWith('[[')) {
      line.type = 'note';
    } else if (trimmed.startsWith('#')) {
      line.type = 'section'; expectingDialogue = false; activeCharacter = '';
    } else if (trimmed.startsWith('=') && !trimmed.startsWith('==')) {
      line.type = 'synopsis'; expectingDialogue = false; activeCharacter = '';
    } else if (trimmed.startsWith('~') && !trimmed.startsWith('\\~')) {
      line.type = 'lyric'; line.character = activeCharacter || undefined;
      if (activeCharacter) {
        const stats = characterMap.get(activeCharacter); const words = countWords(printableText(line));
        if (stats) { stats.dialogueLines++; stats.dialogueWords += words; document.wordCount += words; }
      }
    } else if (trimmed.startsWith('>') && trimmed.endsWith('<')) {
      line.type = 'centered'; expectingDialogue = false; activeCharacter = '';
    } else {
      const forcedScene = trimmed.startsWith('.') && !trimmed.startsWith('..');
      const forcedCharacter = trimmed.startsWith('@') && !trimmed.startsWith('\\@');
      const forcedAction = trimmed.startsWith('!') && !trimmed.startsWith('\\!');
      const forcedTransition = trimmed.startsWith('>') && !trimmed.endsWith('<');
      const comparison = forcedScene || forcedCharacter || forcedAction || forcedTransition ? trimmed.slice(1).trimStart() : trimmed;

      if (forcedScene || scenePattern.test(comparison)) {
        line.type = 'scene'; line.forced = forcedScene; line.sceneNumber = comparison.match(/\s+#([^#]+)#\s*$/)?.[1]; document.sceneCount++; currentScene = document.sceneCount; expectingDialogue = false; activeCharacter = '';
      } else if (forcedTransition || isTransition(comparison)) {
        line.type = 'transition'; line.forced = forcedTransition; expectingDialogue = false; activeCharacter = '';
      } else if (forcedCharacter || isNaturalCharacter(comparison, previousBlank, nextLine)) {
        line.type = 'character'; line.forced = forcedCharacter; activeCharacter = normalizeCharacter(comparison); activeDual = /\^\s*$/.test(comparison); line.character = activeCharacter; line.dualDialogue = activeDual; expectingDialogue = true;
        if (!characterMap.has(activeCharacter)) characterMap.set(activeCharacter, { name: activeCharacter, dialogueLines: 0, dialogueWords: 0, sceneCount: 0, estimatedSeconds: 0 });
        if (currentScene >= 0) { if (!characterScenes.has(activeCharacter)) characterScenes.set(activeCharacter, new Set()); characterScenes.get(activeCharacter)!.add(currentScene); }
      } else if (expectingDialogue && trimmed.startsWith('(') && trimmed.endsWith(')')) {
        line.type = 'parenthetical'; line.character = activeCharacter; line.dualDialogue = activeDual;
      } else if (expectingDialogue) {
        line.type = 'dialogue'; line.character = activeCharacter; line.dualDialogue = activeDual;
        const words = countWords(printableText(line)); document.wordCount += words;
        const stats = characterMap.get(activeCharacter); if (stats) { stats.dialogueLines++; stats.dialogueWords += words; }
      } else {
        line.type = 'action'; line.forced = forcedAction; document.wordCount += countWords(printableText(line));
      }
    }

    const opens = (lineText.match(/(?<!\\)\[\[/g) || []).length;
    const closes = (lineText.match(/(?<!\\)\]\]/g) || []).length;
    noteDepth += opens - closes;
    if (line.type === 'action' && likelyScenePattern.test(trimmed)) {
      let replacement = trimmed.toUpperCase();
      if (replacement.startsWith('IN ')) replacement = `INT.${replacement.slice(2)}`;
      else if (replacement.startsWith('EX ')) replacement = `EXT.${replacement.slice(2)}`;
      document.diagnostics.push({ line: index, start: offset, length: lineText.length, message: 'Possible scene heading typo or missing INT./EXT. prefix.', replacement });
    }
    if (line.type === 'transition' && !line.forced && !trimmed.endsWith(':'))
      document.diagnostics.push({ line: index, start: offset, length: lineText.length, message: 'Transitions conventionally end with a colon.', replacement: `${trimmed}:` });
    document.lines.push(line); offset += lineText.length + 1;
  });

  if (inBoneyard) document.diagnostics.push({ line: Math.max(0, rawLines.length - 1), start: text.length, length: 0, message: 'A boneyard is missing its closing */.', replacement: '\n*/' });
  if (noteDepth > 0) document.diagnostics.push({ line: Math.max(0, rawLines.length - 1), start: text.length, length: 0, message: 'A note is missing its closing ]].', replacement: ']]' });
  document.characters = [...characterMap.values()].map((stats) => {
    stats.sceneCount = characterScenes.get(stats.name)?.size ?? 0;
    stats.estimatedSeconds = stats.dialogueWords * 0.38 + stats.sceneCount * 8;
    return stats;
  }).sort((a, b) => b.estimatedSeconds - a.estimatedSeconds || a.name.localeCompare(b.name));
  return document;
}

export function estimatedRuntime(document: FountainDocument): string {
  const dialogueWords = document.characters.reduce((sum, character) => sum + character.dialogueWords, 0);
  const seconds = dialogueWords * 0.38 + Math.max(0, document.wordCount - dialogueWords) * 0.3 + document.sceneCount * 8;
  const minutes = seconds / 60;
  return minutes < 1 ? '<1 min' : minutes < 10 ? `${minutes.toFixed(1)} min` : `${Math.round(minutes)} min`;
}

export function lineTypeLabel(type: LineType) {
  return type.replace('-', ' ').replace(/^./, (letter) => letter.toUpperCase());
}
